test/core/utils/qsa.js

```diff
@@ -57,190 +57,240 @@ describe('axe.utils.querySelectorAllFilter', function() {
+          assert.isUndefined(dom[0]._selectorCache);
+        });
+      } else {
+        beforeEach(function() {
+          dom = getTestDom();
+
+          // prove we're using the cache by deleting all the children
+          dom[0].children = [];
+        });
+
+        it('should not use the cache if not using the top-level node', function() {
+          var nodes = axe.utils.querySelectorAllFilter(dom, 'ul');
+
+          // this would return 4 nodes if we were still using the
+          // top-level cache
+          var result = axe.utils.querySelectorAllFilter(nodes[0], 'li');
+          assert.equal(result.length, 2);
+        });
+      }
+
+      it('should find nodes using just the tag', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'li');
+        assert.equal(result.length, 4);
+      });
+      it('should find nodes using parent selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'ul > li');
+        assert.equal(result.length, 4);
+      });
+      it('should NOT find nodes using parent selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'div > li');
+        assert.equal(result.length, 0);
+      });
+      it('should find nodes using nested parent selectors', function() {
+        var result = axe.utils.querySelectorAllFilter(
+          dom,
+          'span > span > span > span'
+        );
+        assert.equal(result.length, 2);
+      });
+      it('should find nodes using hierarchical selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'div li');
+        assert.equal(result.length, 4);
+      });
+      it('should find nodes using class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.breaking');
+        assert.equal(result.length, 2);
+      });
+      it('should find nodes using hierarchical class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.first .breaking');
+        assert.equal(result.length, 2);
+      });
+      it('should NOT find nodes using hierarchical class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.second .breaking');
+        assert.equal(result.length, 0);
+      });
+      it('should find nodes using multiple class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.second.third');
+        assert.equal(result.length, 1);
+      });
+      it('should find nodes using id', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '#one');
+        assert.equal(result.length, 1);
+      });
+
+      // can only select shadow dom nodes when we're not using the
+      // top-level node. but since the top-level node is the one
+      // with the cache, this only works when we are testing the full
+      // tree (i.e. without cache)
+      if (describeName === 'without cache') {
```

Comment:
This wrapping `if` is also new to handle the test for shadowDOM id querying (which only works when not using the cache for this particular test).